File: JavaScript/1740-Find Distance in a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} p
 * @param {number} q
 * @return {number}
 */
var findDistance = function(root, p, q) {
// find LCA, then BFS
    let LCA = (node) =>{
        if (!node) return null;
        if (node.val === p || node.val === q) return node;
        let left = LCA(node.left), right = LCA(node.right);
        if (left && right){
            return node;
        }else{
            return (left || right);
        }
    }
    
    let LCAnode = LCA(root);
    
    let findDist = (ancestor, value) =>{
        let queue = [[ancestor, 0]];
        while (queue.length){
            size = queue.length;
            for (let i=0; i<size; i++){
                let [curr, step] = queue.shift();
                if (curr.val === value){
                    return step;
                }
                if (curr.left){
                    queue.push([curr.left, step+1]);
                }
                if (curr.right){
                    queue.push([curr.right, step+1]);
                }
            }
        }
    }
    
    return findDist(LCAnode, p) + findDist(LCAnode, q);
};

File: C++/1740-Find Distance in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* getLCA(TreeNode* node, int p, int q) {
        if (node == nullptr) return nullptr;
        if (node->val == p || node->val == q) return node;
        
        TreeNode* node1 = getLCA(node->left, p, q);
        TreeNode* node2 = getLCA(node->right, p, q);

        if (node1 != nullptr && node2 != nullptr) return node;
        else return (node1 == nullptr)?node2:node1;
    }
    int getDistance(TreeNode* node, int target) {
        if (node == nullptr) return INT_MAX/2;
        if (node->val == target) return 0;

        return 1 + min(getDistance(node->left, target), getDistance(node->right, target));
    }
    int findDistance(TreeNode* root, int p, int q) {
        TreeNode* LCA = getLCA(root, p, q);

        int pd = getDistance(LCA, p), qd = getDistance(LCA, q);

        return pd + qd;
    }
};